File: spec/game/combat_feedback_spec.lua
local combat_feedback = require("game.presentation.combat_feedback")
local combat_presentation = require("game.presentation.combat")
local loadouts = require("data.loadouts")
local match_snapshot = require("sim.match_snapshot")
local combat = require("sim.combat")
local match = require("sim.match")
local teams = require("data.teams")
local t = require("spec.support.runner")

---@param kind CombatEventKind
---@param result CombatContactResult?
---@return CombatEvent
local function event(kind, result)
    return {
        kind = kind,
        tick = 12,
        family_id = kind == "guard_recoil" and "guard" or "light_melee",
        source_index = 2,
        target_index = kind == "commit" or kind == "projectile_spawn" and nil or 7,
        source_sequence = 4,
        result = result,
        x = 320,
        y = 240,
    }
end

t.describe("combat feedback presentation contract", function()
    t.it(
        "gives every supported authoritative result a deliberate multimodal disposition",
        function()
            local fixtures = {
                event("commit"),
                event("projectile_spawn"),
                event("projectile_expire"),
                event("contact", "hit"),
                event("contact", "extended"),
                event("contact", "guarded"),
                event("contact", "immune"),
                event("contact", "superseded"),
                event("ball_spill"),
                event("forced"),
                event("guard_recoil"),
            }
            local semantic_ids = {}
            for index, value in ipairs(fixtures) do
                local link = combat_feedback.accepted(value, "accepted/" .. index)
                t.eq(link.link_kind, "accepted_encounter")
                t.eq(link.source_sequence, 4)
                t.eq(link.feedback_event_id, "accepted/" .. index)
                t.is_true(link.disposition.vfx ~= "none")
                t.is_true(link.disposition.audio ~= nil)
                t.is_true(link.disposition.glyph ~= "")
                semantic_ids[link.semantic_id] = true
            end
            t.eq(#fixtures, 11)
            local semantic_count = 0
            for _ in pairs(semantic_ids) do
                semantic_count = semantic_count + 1
            end
            t.eq(semantic_count, 11)
        end
    )

    t.it("keeps accepted, rejected, and missing-feedback linkage disjoint", function()
        local accepted = combat_feedback.accepted(event("contact", "hit"), "accepted/contact")
        local rejected =
            combat_feedback.rejected_request("request/rejected", 14, 2, "soccer_commitment")
        local missing =
            combat_feedback.missing_feedback("request/missing", 15, 2, "missing_press_edge")

        t.eq(accepted.link_kind, "accepted_encounter")
        t.eq(accepted.source_sequence, 4)
        t.eq(rejected.link_kind, "rejected_request")
        t.is_true(rejected.source_sequence == nil)
        t.eq(rejected.reason, "soccer_commitment")
        t.eq(rejected.feedback_event_id, "request/rejected")
        t.eq(rejected.disposition.vfx, "rejected")
        t.eq(missing.link_kind, "missing_feedback")
        t.is_true(missing.source_sequence == nil)
        t.is_true(missing.feedback_event_id == nil)
        t.eq(missing.disposition.vfx, "none")
        t.is_true(missing.disposition.audio == nil)
    end)

    t.it("deduplicates confirmed HUD/camera feedback and bounds reduced-motion behavior", function()
        local state = combat_feedback.new()
        local link = combat_feedback.accepted(event("contact", "hit"), "confirmed/hit")
        t.is_true(combat_feedback.confirm(state, link))
        t.is_true(not combat_feedback.confirm(state, link))
        t.eq(combat_feedback.diagnostics(state).confirmed_count, 1)
        t.eq(assert(combat_feedback.notice(state, 2)).text, "CLEAN HIT")
        t.is_true(combat_feedback.notice(state, 5) == nil)
        local offset = combat_feedback.camera_offset(state)
        t.is_true(math.abs(offset.x) <= 4 and math.abs(offset.y) <= 4)

        local observation = combat_feedback.observation(state, link, "confirmed")
        t.is_true(observation.camera_enabled)
        t.eq(observation.cue_state, "confirmed")

        combat_feedback.configure(state, true, true)
        offset = combat_feedback.camera_offset(state)
        t.eq(offset.x, 0)
        t.eq(offset.y, 0)
        observation = combat_feedback.observation(state, link, "confirmed")
        t.is_true(not observation.camera_enabled)
        t.is_true(observation.reduced_motion)
        t.is_true(observation.reduced_flash)

        combat_feedback.tick(state, 1)
        t.is_true(combat_feedback.notice(state, 2) == nil)
    end)

    t.it("never mutates simulation truth while projecting feedback", function()
        local state = match.new({
            home = teams.nebula,
            away = teams.orion,
            field = { w = 960, h = 540 },
        })
        local combat_state = combat.new_state(state)
        local before = match_snapshot.hash(match_snapshot.capture(state, combat_state))
        local link = combat_feedback.accepted(event("contact", "guarded"), "pure/guarded")
        local feedback_state = combat_feedback.new()
        combat_feedback.confirm(feedback_state, link)
        combat_feedback.tick(feedback_state, 1 / 60)
        combat_feedback.observation(feedback_state, link, "confirmed")
        local after = match_snapshot.hash(match_snapshot.capture(state, combat_state))
        t.eq(before, after)
    end)

    t.it("keeps simulation identity stable across a same-family presentation swap", function()
        local state = match.new({
            home = teams.nebula,
            away = teams.orion,
            field = { w = 960, h = 540 },
        })
        local combat_state = combat.new_state(state)
        local player_index
        for index, runtime in ipairs(combat_state.players) do
            if runtime.family_id == "light_melee" then
                player_index = index
                break
            end
        end
        local index = assert(player_index)
        local runtime = combat_state.players[index]
        local loadout = assert(loadouts[assert(runtime.loadout_id)])
        local original = loadout.equipment_presentation_id
        local swapped = original == "toy_foam_sword" and "scifi_energy_blade" or "toy_foam_sword"
        local before_hash = match_snapshot.hash(match_snapshot.capture(state, combat_state))
        local before_name =
            assert(combat_presentation.model(state, combat_state).players[index].equipment_name)
        local ok, swapped_name = pcall(function()
            loadout.equipment_presentation_id = swapped
            return assert(
                combat_presentation.model(state, combat_state).players[index].equipment_name
            )
        end)
        loadout.equipment_presentation_id = original
        assert(ok, swapped_name)
        local after_hash = match_snapshot.hash(match_snapshot.capture(state, combat_state))
        t.is_true(swapped_name ~= before_name)
        t.eq(after_hash, before_hash)
    end)
end)

File: ts/packages/presentation/src/combat_feedback.spec.ts
// Fixture note: `combat_feedback`'s functions never take a `MatchState`/
// `CombatMatchState` as an argument in the first place — they operate only
// on detached `CombatEvent` records and this module's own
// `CombatFeedbackState`. So "never mutates simulation truth" is checked
// with an explicit structural before/after snapshot of a synthetic
// sim-shaped fixture, rather than a hash comparison.
//
// The "same-family presentation swap" test passes two different data
// objects rather than mutating one `CombatPresentationData` in place and
// restoring it afterward. Since `combat.model` takes its data tables as an
// explicit parameter (see combat.ts's boundary note) rather than a
// captured global, this makes "a cosmetic swap cannot touch sim state" true
// by construction rather than by convention.

import { describe, expect, it } from "vitest";
import { Vec2 } from "@gc/core";
import {
  combat,
  type CombatContactResult,
  type CombatEvent,
  type CombatEventKind,
  type CombatMatchState,
  type CombatPresentationData,
  type MatchState,
} from "./combat.ts";
import { combatFeedback } from "./combat_feedback.ts";

function event(kind: CombatEventKind, result?: CombatContactResult): CombatEvent {
  return {
    kind,
    tick: 12,
    family_id: kind === "guard_recoil" ? "guard" : "light_melee",
    source_index: 2,
    // `target_index` is intentionally omitted here for "commit" and
    // "projectile_spawn": no assertion in this spec reads it for those
    // kinds, so this fixture uses the clearly-intended value (absent)
    // rather than a stray placeholder in a `number?` field.
    ...(kind !== "commit" && kind !== "projectile_spawn" ? { target_index: 7 } : {}),
    source_sequence: 4,
    ...(result !== undefined ? { result } : {}),
    x: 320,
    y: 240,
  };
}

describe("combat feedback presentation contract", () => {
  it("gives every supported authoritative result a deliberate multimodal disposition", () => {
    const fixtures: CombatEvent[] = [
      event("commit"),
      event("projectile_spawn"),
      event("projectile_expire"),
      event("contact", "hit"),
      event("contact", "extended"),
      event("contact", "guarded"),
      event("contact", "immune"),
      event("contact", "superseded"),
      event("ball_spill"),
      event("forced"),
      event("guard_recoil"),
    ];
    const semanticIds = new Set<string>();
    fixtures.forEach((value, index) => {
      const link = combatFeedback.accepted(value, `accepted/${index + 1}`);
      expect(link.link_kind).toBe("accepted_encounter");
      expect(link.source_sequence).toBe(4);
      expect(link.feedback_event_id).toBe(`accepted/${index + 1}`);
      expect(link.disposition.vfx).not.toBe("none");
      expect(link.disposition.audio).not.toBeUndefined();
      expect(link.disposition.glyph).not.toBe("");
      semanticIds.add(link.semantic_id);
    });
    expect(fixtures).toHaveLength(11);
    expect(semanticIds.size).toBe(11);
  });

  it("keeps accepted, rejected, and missing-feedback linkage disjoint", () => {
    const accepted = combatFeedback.accepted(event("contact", "hit"), "accepted/contact");
    const rejected = combatFeedback.rejectedRequest("request/rejected", 14, 2, "soccer_commitment");
    const missing = combatFeedback.missingFeedback("request/missing", 15, 2, "missing_press_edge");

    expect(accepted.link_kind).toBe("accepted_encounter");
    expect(accepted.source_sequence).toBe(4);
    expect(rejected.link_kind).toBe("rejected_request");
    expect(rejected.source_sequence).toBeUndefined();
    expect(rejected.reason).toBe("soccer_commitment");
    expect(rejected.feedback_event_id).toBe("request/rejected");
    expect(rejected.disposition.vfx).toBe("rejected");
    expect(missing.link_kind).toBe("missing_feedback");
    expect(missing.source_sequence).toBeUndefined();
    expect(missing.feedback_event_id).toBeUndefined();
    expect(missing.disposition.vfx).toBe("none");
    expect(missing.disposition.audio).toBeUndefined();
  });

  it("deduplicates confirmed HUD/camera feedback and bounds reduced-motion behavior", () => {
    const state = combatFeedback.new();
    const link = combatFeedback.accepted(event("contact", "hit"), "confirmed/hit");
    expect(combatFeedback.confirm(state, link)).toBe(true);
    expect(combatFeedback.confirm(state, link)).toBe(false);
    expect(combatFeedback.diagnostics(state).confirmed_count).toBe(1);
    expect(combatFeedback.notice(state, 2)?.text).toBe("CLEAN HIT");
    expect(combatFeedback.notice(state, 5)).toBeNull();
    const offset = combatFeedback.cameraOffset(state);
    expect(Math.abs(offset.x)).toBeLessThanOrEqual(4);
    expect(Math.abs(offset.y)).toBeLessThanOrEqual(4);

    let observation = combatFeedback.observation(state, link, "confirmed");
    expect(observation.camera_enabled).toBe(true);
    expect(observation.cue_state).toBe("confirmed");

    combatFeedback.configure(state, true, true);
    const reducedOffset = combatFeedback.cameraOffset(state);
    expect(reducedOffset.x).toBe(0);
    expect(reducedOffset.y).toBe(0);
    observation = combatFeedback.observation(state, link, "confirmed");
    expect(observation.camera_enabled).toBe(false);
    expect(observation.reduced_motion).toBe(true);
    expect(observation.reduced_flash).toBe(true);

    combatFeedback.tick(state, 1);
    expect(combatFeedback.notice(state, 2)).toBeNull();
  });

  it("never mutates simulation truth while projecting feedback", () => {
    // See file header: this module's functions never take sim state as an
    // argument, so a structural snapshot substitutes for a hash comparison
    // here.
    const state: MatchState = {
      players: [{ id: "home_mid_1", pos: new Vec2(100, 200), facing: new Vec2(1, 0) }],
    };
    const combatState: CombatMatchState = {
      tick: 0,
      player_ids: ["home_mid_1"],
      players: [
        {
          phase: "ready",
          phase_ticks: 0,
          cooldown_ticks: 0,
          forced_ticks: 0,
          immunity_ticks: 0,
          family_id: "light_melee",
          loadout_id: "loadout_tournament_sword",
        },
      ],
      projectiles: [],
    };
    const before = JSON.stringify({ state, combatState });

    const link = combatFeedback.accepted(event("contact", "guarded"), "pure/guarded");
    const feedbackState = combatFeedback.new();
    combatFeedback.confirm(feedbackState, link);
    combatFeedback.tick(feedbackState, 1 / 60);
    combatFeedback.observation(feedbackState, link, "confirmed");

    const after = JSON.stringify({ state, combatState });
    expect(after).toBe(before);
  });

  it("keeps simulation identity stable across a same-family presentation swap", () => {
    const state: MatchState = {
      players: [{ id: "home_mid_1", pos: new Vec2(100, 200), facing: new Vec2(1, 0) }],
    };
    const combatState: CombatMatchState = {
      tick: 0,
      player_ids: ["home_mid_1"],
      players: [
        {
          phase: "ready",
          phase_ticks: 0,
          cooldown_ticks: 0,
          forced_ticks: 0,
          immunity_ticks: 0,
          family_id: "light_melee",
          loadout_id: "loadout_light_melee_test",
        },
      ],
      projectiles: [],
    };
    const before = JSON.stringify({ state, combatState });

    function dataWith(equipmentPresentationId: string): CombatPresentationData {
      return {
        action_families: {
          unarmed: {
            id: "unarmed",
            name: "Unarmed",
            windup_ticks: 6,
            active_ticks: 4,
            recovery_ticks: 12,
            cooldown_ticks: 24,
            reach_px: 30,
            front_arc_degrees: 100,
          },
          guard: { id: "guard", name: "Guard", windup_ticks: 6, recovery_ticks: 9, cooldown_ticks: 0, front_arc_degrees: 120 },
          light_melee: {
            id: "light_melee",
            name: "Light Melee",
            windup_ticks: 12,
            active_ticks: 5,
            recovery_ticks: 21,
            cooldown_ticks: 42,
            reach_px: 42,
            front_arc_degrees: 75,
          },
          ranged: {
            id: "ranged",
            name: "Ranged",
            windup_ticks: 18,
            active_ticks: 1,
            recovery_ticks: 27,
            cooldown_ticks: 60,
            projectile_speed_px_per_second: 300,
            projectile_lifetime_ticks: 60,
            front_arc_degrees: 20,
          },
        },
        equipment_presentations: {
          toy_foam_sword: { id: "toy_foam_sword", name: "Foam Champion", family_id: "light_melee", attachment: "right_hand" },
          scifi_energy_blade: {
            id: "scifi_energy_blade",
            name: "Vector Blade",
            family_id: "light_melee",
            attachment: "right_hand",
          },
        },
        loadouts: {
          loadout_light_melee_test: {
            id: "loadout_light_melee_test",
            family_id: "light_melee",
            equipment_presentation_id: equipmentPresentationId,
          },
        },
      };
    }

    const beforeName = combat.model(state, combatState, dataWith("toy_foam_sword")).players[0]?.equipment_name;
    expect(beforeName).toBeDefined();
    const swappedName = combat.model(state, combatState, dataWith("scifi_energy_blade")).players[0]?.equipment_name;
    expect(swappedName).toBeDefined();

    const after = JSON.stringify({ state, combatState });
    expect(swappedName).not.toBe(beforeName);
    expect(after).toBe(before);
  });
});
